bot/utils/blurple_formatter.py

```diff
@@ -0,0 +1,732 @@
+        ast.Constant,
+        ast.List,
+        ast.ListComp,
+        ast.Dict,
+        ast.DictComp,
+        ast.Set,
+        ast.SetComp,
+        ast.Tuple,
+        ast.GeneratorExp,
+    },
+    {ast.Subscript, ast.Call, ast.Attribute},
+    {ast.Await},
+    {ast.Pow},
+    {ast.UAdd, ast.USub, ast.Invert},
+    {ast.Mult, ast.MatMult, ast.Div, ast.FloorDiv, ast.Mod},
+    {ast.Add, ast.Sub},
+    {ast.LShift, ast.RShift},
+    {ast.BitAnd},
+    {ast.BitXor},
+    {ast.BitOr},
+    {ast.BinOp},
+    {
+        ast.Compare,
+        ast.In,
+        ast.NotIn,
+        ast.Is,
+        ast.IsNot,
+        ast.Lt,
+        ast.LtE,
+        ast.Gt,
+        ast.GtE,
+        ast.NotEq,
+        ast.Eq,
+    },
+    {ast.Not},
+    {ast.And},
+    {ast.Or},
+    {ast.IfExp},
+    {ast.Lambda},
+    {ast.NamedExpr},
+]
+precedences = {op: i for i, ops in enumerate(precedences[::-1]) for op in ops}
+
+
+# right associative operators
+r_assoc = {ast.Pow}
+
+# statements which can have a semicolon after them
+semicolon_able = (
+    ast.AnnAssign,
+    ast.Assert,
+    ast.Assign,
+    ast.AugAssign,
+    ast.Break,
+    ast.Continue,
+    ast.Delete,
+    ast.Expr,
+    ast.Global,
+    ast.Import,
+    ast.ImportFrom,
+    ast.Nonlocal,
+    ast.Pass,
+    ast.Raise,
+    ast.Return,
+)
+
+# operators which the newline goes before
+nl_before_op = {ast.Add, ast.Mult}
+
+
+def space(a: int = 0, b: int = 3) -> str:
+    """Randomly generate whitespace of length between the given bounds."""
+    return " " * randrange(a, b)
+
+
+def num_spaces(src: str) -> int:
+    """Count the number of spaces at the beginning of the string."""
+    return len(src) - len(src.lstrip())
+
+
+def invert_indents(src: str) -> str:
+    """
+    Invert the indentation on each line of the given string.
+
+    Lines with the maximum indentation become least indented, and vice versa.
+    """
+    lines = src.splitlines()
+    max_spaces = max(num_spaces(line) for line in lines)
+    return "\n".join(
+        (max_spaces - num_spaces(line)) * " " + line.lstrip() for line in lines
+    )
+
+
+def parenthesize(node: ast.AST, _nl_able: bool = False) -> str:
+    """Wrap the un-parsed node in parentheses."""
+    return f"({unparse(node, True)})"
+
+
+def indent(nodes: list[ast.AST], n: int) -> Generator[str, None, None]:
+    """
+    Indent each line of the un-parsed nodes by the given number of spaces.
+
+    Multiple nodes are put on the same line, separated by semicolons, if possible
+    until they hit the line length limit.
+    """
+    curr_chunk = []
+    curr_len = 0
+    for stmt in nodes:
+        stmt_lines = unparse(stmt).splitlines()
+        if isinstance(stmt, semicolon_able):
+            for line in stmt_lines:
+                curr_chunk.append(line)
+                curr_len += len(line)
+                if curr_len > max_line_length or curr_chunk[-1][-1] != ";":
+                    yield f"{n * ' '}{f'{space()}'.join(curr_chunk)}"
```

Comment:
Just use concatenation here!

```suggestion
                    yield n * " " + space().join(curr_chunk)}
```

Reply:
resolved in 8c224a52be5000af380465a9d71a55997750fccd